doc/source/data/dataset.rst

```diff
@@ -377,7 +377,178 @@ Tensor datasets can also be created from NumPy ndarrays that are already stored
+
+With our Pandas extension type, :class:`TensorDtype <ray.data.extensions.tensor_extension.TensorDtype>`, and extension array, :class:`TensorArray <ray.data.extensions.tensor_extension.TensorArray>`, you can do familiar aggregations and arithmetic, comparison, and logical operations on a DataFrame containing a tensor column and the operations will be applied to the underlying tensors as expected. With our Arrow extension type, :class:`ArrowTensorType <ray.data.extensions.tensor_extension.ArrowTensorType>`, and extension array, :class:`ArrowTensorArray <ray.data.extensions.tensor_extension.ArrowTensorArray>`, you'll be able to import that DataFrame into Ray Datasets and read/write the data from/to the Parquet format.
+
+Automatic conversion between the Pandas and Arrow extension types/arrays keeps the details under-the-hood, so you only have to worry about casting the column to a tensor column using our Pandas extension type when first ingesting the table into a ``Dataset``, whether from storage or in-memory. All table operations downstream from that cast should work automatically.
+
+If you already have Parquet files with tensor columns containing serialized tensors (e.g. raw or pickle bytes), you can manually cast this tensor column to our tensor extension type via a read-time user-defined function. This UDF will be pushed down to Ray Datasets' IO layer and executed on each block in parallel, as it's read from storage.
+
+.. code-block:: python
+    import pickle
+    import numpy as np
+    import pandas as pd
+    from ray.data.extensions import TensorDtype, TensorArray
+
+    path = "/tmp/some_path"
+
+    # Create a DataFrame with a list of pickled ndarrays as a column.
+    # Note that we do not cast it to a tensor array, so each element in the
+    # column is an opaque blob of bytes.
+    arr = np.arange(24).reshape((3, 2, 2, 2))
+    df = pd.DataFrame({
+        "one": [1, 2, 3],
+        "two": [pickle.dumps(tensor) for tensor in arr]})
+
+    # Write the dataset to Parquet. The tensor column will be written as an
+    # array of opaque byte blobs.
+    ds = ray.data.from_pandas([ray.put(df)])
+    ds.write_parquet(path)
+
+    # Manually deserialize the tensor pickle bytes and cast to our tensor
+    # extension type. For the sake of efficiency, we directly construct a
+    # TensorArray rather than .astype() casting on the mutated column with
+    # TensorDtype.
+    def cast_udf_direct(block: pa.Table) -> pa.Table:
+        block = block.to_pandas()
+        block["two"] = TensorArray([pickle.loads(a) for a in block["two"]])
+        return pa.Table.from_pandas(block)
+
+    # Read the Parquet files into a new Dataset, applying the casting UDF
+    # on-the-fly within the underlying read tasks.
+    ds = ray.data.read_parquet(path, block_udf=cast_udf)
```

Comment:
A couple comments here:
1. Can we keep block_udf a private API for now?
2. The user experience for reading bytes columns as tensors should be really good. Suppose column "f2" can be interpreted as a tensor array, why not just support something like `ray.data.read_parquet(..., tensor_schema={"f2": (np.float32, (30, 30, 2))})` and have that generate the right udf under the hood?

Keep in mind the *primary entrypoint* of this feature will be reading existing data as tensor type, not writing. Hence, we should optimize our APIs for reads.

Reply:
Btw, we can probably implement map pushdown automatically as opportunistic on reads.

For example, if you did `ray.data.read_parquet().map(f)`, we can fuse `f` with the read tasks. But this is a class of optimizations for a later date.

Reply:
1. We can keep it a private API, i.e. change the parameter to `_block_udf`. I can make that change.
2. I was hoping to hide this stopgap solution behind a generically useful API like map pushdown so we don't have a clunky deprecation cycle in a few months once extension type casting kernels are supported, but agreed that it's better UX, which should be the highest priority. I'll add a private `_tensor_schema` parameter that can be used instead and update the docs to use that casting method.

>Btw, we can probably implement map pushdown automatically as opportunistic on reads.

>For example, if you did ray.data.read_parquet().map(f), we can fuse f with the read tasks. But this is a class of optimizations for a later date.

Of course, I stayed away from automatic task fusion here to keep the PR small and since we probably want a design doc around enabling such optimizations, but ideally we'd automatically fuse `lazy_ingest().map().filter()` into the read tasks. We might want to allocate some time for creating such a design doc at the beginning of Q4 if we think these optimizations will be high impact.

Reply:
For the `_tensor_schema` parameter, we have a few options:

1. Arrow-agnostic dict of column name --> NumPy dtype and shape tuple: `_tensor_schema = {"f2": (np.float32, (2, 30, 30))}`
2. Arrow-aware dict of column name --> Arrow dtype and shape tuple: `_tensor_schema = {"f2": (pa.float32(), (2, 30, 30))}`
3. Arrow schema using our extension type: `_tensor_schema = pa.schema([("f2", ArrowTensorType(pa.float32(), (2, 30 30)))])`

(1) requires no knowledge of Arrow and probably provides the best UX overall, but I think that (3) aligns best with an Arrow-centric API, which might be what we want for Parquet if we're going to expose e.g. pushdown filter expressions, and aligns best with our current implementation (doing it manually via a cast in the read task) and future implementation (doing it automatically via a custom casting kernel for our extension type).

If we're planning on iterating on this anyway, I'd say that we should go with (1) first since it provides the best UX, and we can reexamine later if our Parquet reading API gets more Arrow-centric, or once we're looking at stabilizing this API.

Reply:
Agree (1) is probably easiest for now. I do like how (3) is much more Arrow native, maybe it can also be useful for overriding the schema of other fields as well in the future.